flink-yarn/src/main/java/org/apache/flink/yarn/ApplicationReportProviderImpl.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.yarn;
+
+import org.apache.flink.annotation.Internal;
+
+import org.apache.hadoop.yarn.api.records.ApplicationId;
+import org.apache.hadoop.yarn.api.records.ApplicationReport;
+import org.apache.hadoop.yarn.api.records.YarnApplicationState;
+import org.apache.hadoop.yarn.client.api.YarnClient;
+
+import java.io.IOException;
+
+/** Then implementation of ApplicationReportProvider. */
+@Internal
+public class ApplicationReportProviderImpl implements ApplicationReportProvider {
+    private YarnClient yarnClient;
```

Comment:
What is the lifecycle of the `yarnClient`? When will it be freed/closed?

Reply:
The lifecycle of `yarnclient` is the same with YarnClusterDescriptor.

And checked all the invoking `getClusterClient` methods, no one invoke it after YarnClusterDescriptor closed.

Please let me know if i missed something.

Maybe i need to check whether yarnclient is closed and recreate in this case. Do u think so?

Reply:
`yarnclient` is created by `YarnClusterClientFactory.getClusterDescriptor`. But when `YarnClusterDescriptor` is invoked close() method, it will be stopped.

So i think we could make sharedYarnClient true to avoid above problem.